File: packages/sdk/java/packages/core/src/main/java/dev/edgebase/sdk/core/HttpClient.java
// EdgeBase Java SDK — HTTP client.
// OkHttp-based HTTP client with automatic authentication,
// 401 retry, request metadata, and multipart uploads.
package dev.edgebase.sdk.core;

import com.google.gson.Gson;
import okhttp3.*;
import okhttp3.MediaType;

import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.TimeUnit;

/**
 * HTTP client for EdgeBase API communication.
 *
 * <p>
 * Features:
 * <ul>
 * <li>Automatic Bearer token injection</li>
 * <li>401 response → token refresh → automatic retry</li>
 * <li>Project metadata headers</li>
 * <li>Multipart file uploads</li>
 * <li>Public endpoints (no auth required)</li>
 * </ul>
 */
public class HttpClient {
    private static final MediaType JSON_MEDIA_TYPE = MediaType.parse("application/json; charset=utf-8");
    private static final Gson gson = new Gson();

    public final String baseUrl;
    private final TokenManager tokenManager;
    private final String serviceKey;
    private final String projectId;
    private final OkHttpClient client;
    private String locale;

    public HttpClient(String baseUrl, TokenManager tokenManager, ContextManager contextManager) {
        this(baseUrl, tokenManager, contextManager, null, null);
    }

    /**
     * Convenience constructor for service-key-only usage (e.g., E2E tests and admin
     * scripts).
     * Creates no-op TokenManager and ContextManager internally.
     */
    public HttpClient(String baseUrl, String serviceKey) {
        this(baseUrl, new TokenManager() {
            @Override
            public String getAccessToken() {
                return null;
            }

            @Override
            public String getRefreshToken() {
                return null;
            }

            @Override
            public void setTokens(String a, String r) {
            }

            @Override
            public void clearTokens() {
            }
        }, new ContextManager(), serviceKey, null);
    }

    public HttpClient(String baseUrl, TokenManager tokenManager, ContextManager contextManager,
            String serviceKey, String projectId) {
        this.baseUrl = baseUrl;
        this.tokenManager = tokenManager;
        this.serviceKey = serviceKey;
        this.projectId = projectId;
        this.client = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(120, TimeUnit.SECONDS)
                .writeTimeout(120, TimeUnit.SECONDS)
                .build();
    }

    // ─── Public API ───

    @SuppressWarnings("unchecked")
    public Object get(String path) throws EdgeBaseError {
        return get(path, null);
    }

    @SuppressWarnings("unchecked")
    public Object get(String path, Map<String, String> queryParams) throws EdgeBaseError {
        return request("GET", path, null, false, false, queryParams, 0);
    }

    public Object post(String path, Map<String, ?> body) throws EdgeBaseError {
        return request("POST", path, body, false, false, null, 0);
    }

    public Object patch(String path, Map<String, ?> body) throws EdgeBaseError {
        return request("PATCH", path, body, false, false, null, 0);
    }

    public Object put(String path, Map<String, ?> body) throws EdgeBaseError {
        return request("PUT", path, body, false, false, null, 0);
    }

    public Object getWithQuery(String path, Map<String, String> queryParams) throws EdgeBaseError {
        return request("GET", path, null, false, false, queryParams, 0);
    }

    public Object postWithQuery(String path, Map<String, ?> body, Map<String, String> queryParams) throws EdgeBaseError {
        return request("POST", path, body, false, false, queryParams, 0);
    }

    public Object putWithQuery(String path, Map<String, ?> body, Map<String, String> queryParams) throws EdgeBaseError {
        return request("PUT", path, body, false, false, queryParams, 0);
    }

    public Object delete(String path) throws EdgeBaseError {
        return request("DELETE", path, null, false, false, null, 0);
    }

    public Object delete(String path, Map<String, ?> body) throws EdgeBaseError {
        return request("DELETE", path, body, false, false, null, 0);
    }

    public void close() {
        client.dispatcher().executorService().shutdown();
        client.connectionPool().evictAll();
        Cache cache = client.cache();
        if (cache != null) {
            try {
                cache.close();
            } catch (IOException ignored) {
            }
        }
    }

    /** HEAD request — returns true if resource exists (2xx). */
    public boolean head(String path) throws EdgeBaseError {
        try {
            String url = buildUrl(path, null);
            Request.Builder requestBuilder = new Request.Builder().url(url).head();

            if (serviceKey != null) {
                requestBuilder.addHeader("X-EdgeBase-Service-Key", serviceKey);
            } else {
                try {
                    String token = tokenManager.getAccessToken();
                    if (token != null)
                        requestBuilder.addHeader("Authorization", "Bearer " + token);
                } catch (Exception ignored) {
                    // Token refresh failed — proceed as unauthenticated
                }
            }
            addRequestMetadataHeaders(requestBuilder);

            try (Response response = client.newCall(requestBuilder.build()).execute()) {
                return response.isSuccessful();
            }
        } catch (IOException e) {
            throw new EdgeBaseError(0, "HEAD request failed: " + e.getMessage());
        }
    }

    /**
     * POST to public endpoint (no authentication).
     */
    public Object postPublic(String path) throws EdgeBaseError {
        return request("POST", path, Collections.emptyMap(), true, false, null, 0);
    }

    public Object postPublic(String path, Map<String, ?> body) throws EdgeBaseError {
        return request("POST", path, body, true, false, null, 0);
    }

    /**
     * POST raw bytes with optional query params.
     */
    @SuppressWarnings("unchecked")
    public Object postBytes(String path, byte[] data, String contentType, Map<String, String> queryParams) throws EdgeBaseError {
        try {
            String url = buildUrl(path, queryParams);
            RequestBody body = RequestBody.create(data, MediaType.parse(contentType));
            Request.Builder requestBuilder = new Request.Builder().url(url).post(body);

            if (serviceKey != null) {
                requestBuilder.addHeader("X-EdgeBase-Service-Key", serviceKey);
            } else {
                try {
                    String token = tokenManager.getAccessToken();
                    if (token != null)
                        requestBuilder.addHeader("Authorization", "Bearer " + token);
                } catch (Exception ignored) {
                    // Token refresh failed — proceed as unauthenticated
                }
            }
            addRequestMetadataHeaders(requestBuilder);

            try (Response response = client.newCall(requestBuilder.build()).execute()) {
                return parseResponse(response);
            }
        } catch (EdgeBaseError e) {
            throw e;
        } catch (IOException e) {
            throw new EdgeBaseError(0, "Binary POST failed: " + e.getMessage());
        }
    }

    /**
     * Upload file data via multipart POST.
     */
    @SuppressWarnings("unchecked")
    public Object uploadMultipart(String path, String fileName, byte[] data,
            String contentType, Map<String, String> extraFields) throws EdgeBaseError {
        try {
            MultipartBody.Builder multipartBuilder = new MultipartBody.Builder()
                    .setType(MultipartBody.FORM)
                    .addFormDataPart("file", fileName,
                            RequestBody.create(data, MediaType.parse(contentType)));

            if (extraFields != null) {
                for (Map.Entry<String, String> entry : extraFields.entrySet()) {
                    multipartBuilder.addFormDataPart(entry.getKey(), entry.getValue());
                }
            }

            String url = buildUrl(path, null);
            Request.Builder requestBuilder = new Request.Builder().url(url).post(multipartBuilder.build());

            if (serviceKey != null) {
                requestBuilder.addHeader("X-EdgeBase-Service-Key", serviceKey);
            } else {
                try {
                    String token = tokenManager.getAccessToken();
                    if (token != null)
                        requestBuilder.addHeader("Authorization", "Bearer " + token);
                } catch (Exception ignored) {
                    // Token refresh failed — proceed as unauthenticated
                }
            }
            addRequestMetadataHeaders(requestBuilder);

            try (Response response = client.newCall(requestBuilder.build()).execute()) {
                return parseResponse(response);
            }
        } catch (EdgeBaseError e) {
            throw e;
        } catch (IOException e) {
            throw new EdgeBaseError(0, "Upload failed: " + e.getMessage());
        }
    }

    /**
     * Download raw bytes.
     */
    public byte[] downloadRaw(String path) throws EdgeBaseError {
        try {
            String url = buildUrl(path, null);
            Request.Builder requestBuilder = new Request.Builder().url(url).get();

            try {
                String token = tokenManager.getAccessToken();
                if (token != null)
                    requestBuilder.addHeader("Authorization", "Bearer " + token);
            } catch (Exception ignored) {
                // Token refresh failed — proceed as unauthenticated
            }
            addRequestMetadataHeaders(requestBuilder);

            try (Response response = client.newCall(requestBuilder.build()).execute()) {
                if (!response.isSuccessful()) {
                    throw new EdgeBaseError(response.code(), "Download failed: " + response.message());
                }
                ResponseBody body = response.body();
                return body != null ? body.bytes() : new byte[0];
            }
        } catch (EdgeBaseError e) {
            throw e;
        } catch (IOException e) {
            throw new EdgeBaseError(0, "Download failed: " + e.getMessage());
        }
    }

    /**
     * Get the full API base URL (baseUrl + API prefix).
     * Useful for constructing redirect URLs (e.g., OAuth) without hardcoding the /api/ prefix.
     */
    public String getApiBaseUrl() {
        return baseUrl + buildApiPrefix();
    }

    public void setLocale(String locale) {
        this.locale = locale;
    }

    public String getLocale() {
        return locale;
    }

    /**
     * Get underlying OkHttpClient for WebSocket connections.
     */
    public OkHttpClient getOkHttpClient() {
        return client;
    }

    /**
     * Returns a view of this client with a DB path prefix inserted before every API
     * path.
     * Used by {@link DbRef} to scope all table operations to a specific
     * namespace/instance.
     *
     * @param dbPath prefix, e.g. "/db/shared" or "/db/tenant-abc/instance-1"
     */
    public HttpClient withDbPath(String dbPath) {
        // Embed the db path into the baseUrl string so buildUrl appends
        // /api<dbPath>/<tablePath>
        // We reuse the same auth and request metadata machinery.
        return new HttpClient(this.baseUrl + "/api" + dbPath, this.tokenManager,
                new ContextManager(), this.serviceKey, this.projectId) {
            @Override
            protected String buildApiPrefix() {
                return ""; // already embedded in baseUrl
            }
        };
    }

    // ─── Internal ───

    @SuppressWarnings("unchecked")
    private Object request(String method, String path, Map<String, ?> body,
            boolean isPublic, boolean isRetry, Map<String, String> queryParams, int rateLimitAttempt) throws EdgeBaseError {
        try {
            String url = buildUrl(path, queryParams);
            RequestBody requestBody = null;
            if (body != null) {
                requestBody = RequestBody.create(gson.toJson(body), JSON_MEDIA_TYPE);
            }

            Request.Builder requestBuilder = new Request.Builder().url(url);
            switch (method) {
                case "GET":
                    requestBuilder.get();
                    break;
                case "POST":
                    requestBuilder.post(requestBody != null ? requestBody : RequestBody.create("{}", JSON_MEDIA_TYPE));
                    break;
                case "PATCH":
                    requestBuilder.patch(requestBody != null ? requestBody : RequestBody.create("{}", JSON_MEDIA_TYPE));
                    break;
                case "PUT":
                    requestBuilder.put(requestBody != null ? requestBody : RequestBody.create("{}", JSON_MEDIA_TYPE));
                    break;
                case "DELETE":
                    if (requestBody != null) {
                        requestBuilder.delete(requestBody);
                    } else {
                        requestBuilder.delete(RequestBody.create("{}", JSON_MEDIA_TYPE));
                    }
                    break;
            }

            // Add auth headers (skip for public endpoints)
            if (!isPublic) {
                if (serviceKey != null) {
                    requestBuilder.addHeader("X-EdgeBase-Service-Key", serviceKey);
                } else {
                    try {
                        String token = tokenManager.getAccessToken();
                        if (token != null)
                            requestBuilder.addHeader("Authorization", "Bearer " + token);
                    } catch (Exception ignored) {
                        // Token refresh failed — proceed as unauthenticated
                    }
                }
            }

            addRequestMetadataHeaders(requestBuilder);

            // Transport retry: independent loop (max 2 retries) around the network call
            Response response = null;
            for (int transportAttempt = 0; transportAttempt <= 2; transportAttempt++) {
                try {
                    response = client.newCall(requestBuilder.build()).execute();
                    break;
                } catch (IOException e) {
                    if (transportAttempt < 2) {
                        try { Thread.sleep(50L * (transportAttempt + 1)); } catch (InterruptedException ignored) { Thread.currentThread().interrupt(); }
                        // Rebuild request for retry (OkHttp request can only be executed once)
                        requestBuilder = requestBuilder.build().newBuilder();
                        continue;
                    }
                    throw new EdgeBaseError(0, "Request failed: " + e.getMessage());
                }
            }

            try {
                // 429 retry with Retry-After
                if (response.code() == 429 && rateLimitAttempt < 3) {
                    long delay = parseRetryAfterDelay(response.header("Retry-After"), rateLimitAttempt);
                    response.close();
                    Thread.sleep(delay);
                    return request(method, path, body, isPublic, isRetry, queryParams, rateLimitAttempt + 1);
                }

                // Handle 401 — retry once after token refresh
                if (response.code() == 401 && !isRetry && !isPublic) {
                    response.close();
                    String refreshToken = tokenManager.getRefreshToken();
                    if (refreshToken != null) {
                        try {
                            tokenManager.getAccessToken(); // triggers refresh internally
                        } catch (Exception ignored) {
                        }
                    }
                    return request(method, path, body, isPublic, true, queryParams, rateLimitAttempt);
                }
                return parseResponse(response);
            } finally {
                response.close();
            }
        } catch (EdgeBaseError e) {
            throw e;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new EdgeBaseError(0, "Request interrupted");
        }
    }

    private static long parseRetryAfterDelay(String retryAfter, int attempt) {
        long baseDelayMs = 1000L * (1L << attempt);
        if (retryAfter != null) {
            try {
                long seconds = Long.parseLong(retryAfter);
                if (seconds > 0) baseDelayMs = seconds * 1000;
            } catch (NumberFormatException ignored) {}
        }
        long jitter = (long) (baseDelayMs * 0.25 * Math.random());
        return Math.min(baseDelayMs + jitter, 10000);
    }

    protected String buildApiPrefix() {
        return "/api";
    }

    private String buildUrl(String path, Map<String, String> queryParams) {
        StringBuilder sb = new StringBuilder(baseUrl).append(buildApiPrefix()).append(path);
        if (queryParams != null && !queryParams.isEmpty()) {
            sb.append("?");
            boolean first = true;
            for (Map.Entry<String, String> entry : queryParams.entrySet()) {
                if (!first)
                    sb.append("&");
                sb.append(URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8))
                        .append("=")
                        .append(URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8));
                first = false;
            }
        }
        return sb.toString();
    }

    private void addRequestMetadataHeaders(Request.Builder builder) {
        if (projectId != null) {
            builder.addHeader("X-EdgeBase-Project-Id", projectId);
        }
        if (locale != null && !locale.isEmpty()) {
            builder.addHeader("Accept-Language", locale);
        }
    }

    @SuppressWarnings("unchecked")
    private Object parseResponse(Response response) throws EdgeBaseError {
        String bodyStr;
        try {
            ResponseBody body = response.body();
            bodyStr = body != null ? body.string() : "";
        } catch (IOException e) {
            bodyStr = "";
        }

        if (!response.isSuccessful()) {
            try {
                Map<String, Object> parsed = gson.fromJson(bodyStr, Map.class);
                if (parsed != null) {
                    throw EdgeBaseError.fromJson(parsed, response.code());
                }
            } catch (EdgeBaseError e) {
                throw e;
            } catch (Exception ignored) {
            }
            throw new EdgeBaseError(response.code(), bodyStr.isEmpty() ? response.message() : bodyStr);
        }

        if (bodyStr.isEmpty())
            return null;

        String contentType = response.header("Content-Type");
        if (!isLikelyJsonResponse(bodyStr, contentType)) {
            throw new IllegalStateException("Invalid JSON response body");
        }

        try {
            return gson.fromJson(bodyStr, Object.class);
        } catch (Exception e) {
            throw new IllegalStateException("Invalid JSON response body", e);
        }
    }

    private boolean isLikelyJsonResponse(String bodyStr, String contentType) {
        if (contentType != null) {
            String normalized = contentType.toLowerCase(Locale.ROOT);
            if (normalized.contains("/json") || normalized.contains("+json")) {
                return true;
            }
        }

        String trimmed = bodyStr.trim();
        if (trimmed.isEmpty()) {
            return false;
        }
        if (trimmed.startsWith("{") || trimmed.startsWith("[") || trimmed.startsWith("\"") || trimmed.startsWith("-")) {
            return true;
        }
        if (Character.isDigit(trimmed.charAt(0))) {
            return true;
        }
        return trimmed.startsWith("true") || trimmed.startsWith("false") || trimmed.startsWith("null");
    }
}

File: packages/sdk/kotlin/core/src/commonMain/kotlin/dev/edgebase/sdk/core/HttpClient.kt
// EdgeBase Kotlin SDK — HTTP client (Ktor-based, KMP).
//
// Ktor-based HTTP client with automatic authentication,
// 401 retry, and multipart uploads.
//: OkHttp → Ktor for KMP support. #133/#136: ContextManager removed.

package dev.edgebase.sdk.core

import io.ktor.client.HttpClient as KtorHttpClient
import io.ktor.client.request.*
import io.ktor.client.request.forms.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.utils.io.core.*
import kotlinx.coroutines.delay
import kotlinx.serialization.json.*

/**
 * HTTP client for EdgeBase API communication.
 *
 * Features:
 * - Automatic Bearer token injection
 * - 401 response → token refresh → automatic retry
 * - Multipart file uploads
 * - Public endpoints (no auth required)
 */
class HttpClient(
    val baseUrl: String,
    private val tokenManager: TokenManager,
    private val serviceKey: String? = null,
    private val projectId: String? = null,
    private val client: KtorHttpClient = createPlatformHttpClient()
) {
    private val json = Json { ignoreUnknownKeys = true }
    private val maxTransportRetries = 2
    private var locale: String? = null

    // MARK: - Public API

    @Suppress("UNCHECKED_CAST")
    suspend fun get(path: String, queryParams: Map<String, String>? = null): Any? =
        request("GET", path, queryParams = queryParams)

    @Suppress("UNCHECKED_CAST")
    suspend fun post(path: String, body: Map<String, Any?> = emptyMap()): Any? =
        request("POST", path, body)

    /**
     * POST with both JSON body and query parameters.
     * Used by generated core for insert/batch endpoints that accept query params (e.g. upsert=true).
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun postWithQuery(path: String, body: Map<String, Any?> = emptyMap(), queryParams: Map<String, String>? = null): Any? =
        request("POST", path, body, queryParams = queryParams)

    @Suppress("UNCHECKED_CAST")
    suspend fun patch(path: String, body: Map<String, Any?> = emptyMap()): Any? =
        request("PATCH", path, body)

    @Suppress("UNCHECKED_CAST")
    suspend fun put(path: String, body: Map<String, Any?> = emptyMap()): Any? =
        request("PUT", path, body)

    /**
     * PUT with both JSON body and query parameters.
     * Used by generated core for update endpoints that accept query params.
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun putWithQuery(path: String, body: Map<String, Any?> = emptyMap(), queryParams: Map<String, String>? = null): Any? =
        request("PUT", path, body, queryParams = queryParams)

    @Suppress("UNCHECKED_CAST")
    suspend fun delete(path: String, body: Map<String, Any?>? = null): Any? =
        request("DELETE", path, body)

    /** HEAD request — returns true if resource exists (2xx). */
    suspend fun head(path: String): Boolean {
        val url = buildUrl(path)
        val response = executeRequestWithRetry {
            client.request(url) {
                method = HttpMethod.Head
                if (serviceKey != null) {
                    header("X-EdgeBase-Service-Key", serviceKey)
                } else {
                    val token = try { tokenManager.getAccessToken() } catch (_: Exception) { null }
                    if (token != null) header(HttpHeaders.Authorization, "Bearer $token")
                }
                addContextHeaders(this)
            }
        }
        return response.status.isSuccess()
    }

    /**
     * GET from public endpoint (no authentication).
     */
    suspend fun getPublic(path: String, queryParams: Map<String, String>? = null): Any? =
        request("GET", path, queryParams = queryParams, isPublic = true)

    /**
     * POST to public endpoint (no authentication).
     */
    suspend fun postPublic(path: String, body: Map<String, Any?> = emptyMap()): Any? =
        request("POST", path, body, isPublic = true)

    /**
     * POST raw binary data with optional query params.
     */
    suspend fun postBinary(
        path: String,
        data: ByteArray,
        contentType: String = "application/octet-stream",
        queryParams: Map<String, String>? = null,
        isPublic: Boolean = false
    ): Any? {
        val url = buildUrl(path, queryParams)
        val response = executeRequestWithRetry {
            client.request(url) {
                method = HttpMethod.Post
                header(HttpHeaders.ContentType, contentType)
                setBody(data)

                if (!isPublic) {
                    if (serviceKey != null) {
                        header("X-EdgeBase-Service-Key", serviceKey)
                    } else {
                        val token = try { tokenManager.getAccessToken() } catch (_: Exception) { null }
                        if (token != null) header(HttpHeaders.Authorization, "Bearer $token")
                    }
                }

                addContextHeaders(this)
            }
        }

        return parseResponse(response)
    }

    /**
     * Upload file data via multipart POST.
     */
    suspend fun uploadMultipart(
        path: String,
        fileName: String,
        data: ByteArray,
        contentType: String = "application/octet-stream",
        extraFields: Map<String, String> = emptyMap()
    ): Any? {
        val url = buildUrl(path)

        val boundary = "EdgeBaseBoundary${currentTimeMillis()}"
        @Suppress("DEPRECATION")
        val multipartBody = buildPacket {
            extraFields.forEach { (key, value) ->
                writeText("--$boundary\r\n")
                writeText("Content-Disposition: form-data; name=\"$key\"\r\n\r\n")
                writeText(value)
                writeText("\r\n")
            }
            writeText("--$boundary\r\n")
            writeText("Content-Disposition: form-data; name=\"file\"; filename=\"$fileName\"\r\n")
            writeText("Content-Type: $contentType\r\n\r\n")
            writeFully(data)
            writeText("\r\n")
            writeText("--$boundary--\r\n")
        }.readBytes()

        val response = executeRequestWithRetry {
            client.request(url) {
                method = HttpMethod.Post
                header(HttpHeaders.ContentType, "multipart/form-data; boundary=$boundary")
                setBody(multipartBody)

                if (serviceKey != null) {
                    header("X-EdgeBase-Service-Key", serviceKey)
                } else {
                    val token = tokenManager.getAccessToken()
                    if (token != null) header(HttpHeaders.Authorization, "Bearer $token")
                }
                addContextHeaders(this)
            }
        }

        return parseResponse(response)
    }

    /**
     * Download raw bytes.
     */
    suspend fun downloadRaw(path: String): ByteArray {
        val url = buildUrl(path)

        val response = executeRequestWithRetry {
            client.get(url) {
                val token = tokenManager.getAccessToken()
                if (token != null) header(HttpHeaders.Authorization, "Bearer $token")
                addContextHeaders(this)
            }
        }

        if (!response.status.isSuccess()) {
            throw EdgeBaseError(response.status.value, "Download failed: ${response.status.description}")
        }
        return response.bodyAsBytes()
    }

    fun setLocale(locale: String?) {
        this.locale = locale
    }

    fun getLocale(): String? = locale

    // MARK: - Internal

    @Suppress("UNCHECKED_CAST")
    private suspend fun request(
        method: String,
        path: String,
        body: Map<String, Any?>? = null,
        isPublic: Boolean = false,
        isRetry: Boolean = false,
        queryParams: Map<String, String>? = null,
        rateLimitAttempt: Int = 0
    ): Any? {
        val url = buildUrl(path, queryParams)
        val requestBody = body?.let { mapToJsonString(it) }

        val response = executeRequestWithRetry {
            client.request(url) {
                this.method = HttpMethod.parse(method)

                // Set body
                if (requestBody != null) {
                    contentType(ContentType.Application.Json)
                    setBody(requestBody)
                } else if (method != "GET") {
                    contentType(ContentType.Application.Json)
                    setBody("{}")
                }

                // Add auth headers (skip for public endpoints)
                if (!isPublic) {
                    if (serviceKey != null) {
                        header("X-EdgeBase-Service-Key", serviceKey)
                    } else {
                        val token = tokenManager.getAccessToken()
                        if (token != null) header(HttpHeaders.Authorization, "Bearer $token")
                    }
                }

                // Add request metadata headers
                addContextHeaders(this)
            }
        }

        // 429 retry with Retry-After header
        if (response.status.value == 429 && rateLimitAttempt < 3) {
            val retryAfter = response.headers["Retry-After"]
            val baseDelayMs = retryAfter?.toLongOrNull()?.let { it * 1000 }
                ?: (1000L * (1L shl rateLimitAttempt))
            val jitter = (baseDelayMs * 0.25 * kotlin.random.Random.nextDouble()).toLong()
            delay(minOf(baseDelayMs + jitter, 10000L))
            return request(method, path, body, isPublic, isRetry, queryParams, rateLimitAttempt + 1)
        }

        // Handle 401 — retry once after token refresh
        if (response.status.value == 401 && !isRetry && !isPublic) {
            val refreshToken = tokenManager.getRefreshToken()
            if (refreshToken != null) {
                try {
                    tokenManager.getAccessToken() // triggers refresh internally
                } catch (_: Exception) { /* ignore */ }
            }
            return request(method, path, body, isPublic, isRetry = true, queryParams = queryParams)
        }

        return parseResponse(response)
    }

    private suspend fun executeRequestWithRetry(block: suspend () -> HttpResponse): HttpResponse {
        var attempt = 0
        while (true) {
            try {
                return block()
            } catch (error: Throwable) {
                if (attempt >= maxTransportRetries || !isRetryableTransportFailure(error)) {
                    throw error
                }
                attempt += 1
                delay(50L * attempt)
            }
        }
    }

    private fun isRetryableTransportFailure(error: Throwable): Boolean {
        val normalized = (error.message ?: error.toString()).lowercase()
        return normalized.contains("timeout") ||
            normalized.contains("timed out") ||
            normalized.contains("connect") ||
            normalized.contains("connection") ||
            normalized.contains("refused") ||
            normalized.contains("reset") ||
            normalized.contains("econnreset") ||
            normalized.contains("eof") ||
            normalized.contains("end of file") ||
            normalized.contains("prematurely closed") ||
            normalized.contains("closed the connection") ||
            normalized.contains("nxdomain") ||
            normalized.contains("network is unreachable")
    }

    private fun buildUrl(path: String, queryParams: Map<String, String>? = null): String {
        val base = "$baseUrl/api$path"
        if (queryParams.isNullOrEmpty()) return base
        val query = queryParams.entries.joinToString("&") { (k, v) ->
            "${platformUrlEncode(k)}=${platformUrlEncode(v)}"
        }
        return "$base?$query"
    }

    private suspend fun addContextHeaders(builder: HttpRequestBuilder) {
        // #133/#136: X-EdgeBase-Context header removed. namespace+id are in the URL path.
        if (projectId != null) {
            builder.header("X-EdgeBase-Project-Id", projectId)
        }
        if (!locale.isNullOrBlank()) {
            builder.header(HttpHeaders.AcceptLanguage, locale!!)
        }
    }

    @Suppress("UNCHECKED_CAST")
    private suspend fun parseResponse(response: HttpResponse): Any? {
        val bodyStr = response.bodyAsText()

        if (!response.status.isSuccess()) {
            try {
                val parsed = jsonStringToMap(bodyStr)
                throw EdgeBaseError.fromJson(parsed, response.status.value)
            } catch (e: EdgeBaseError) {
                throw e
            } catch (_: Exception) {
                throw EdgeBaseError(response.status.value, bodyStr.ifEmpty { response.status.description })
            }
        }

        if (bodyStr.isEmpty()) return null

        val contentType = response.headers[HttpHeaders.ContentType]
        if (!isLikelyJsonResponse(bodyStr, contentType)) {
            throw IllegalStateException("Invalid JSON response body")
        }

        return try {
            val element = json.parseToJsonElement(bodyStr)
            jsonElementToAny(element)
        } catch (error: Exception) {
            throw IllegalStateException("Invalid JSON response body", error)
        }
    }

    // MARK: - Response body as bytes (Ktor)

    private suspend fun HttpResponse.bodyAsBytes(): ByteArray {
        return this.readRawBytes()
    }

    // MARK: - JSON helpers

    private fun isLikelyJsonResponse(bodyStr: String, contentType: String?): Boolean {
        val normalizedContentType = contentType?.lowercase()
        if (normalizedContentType != null && (normalizedContentType.contains("/json") || normalizedContentType.contains("+json"))) {
            return true
        }

        val trimmed = bodyStr.trim()
        if (trimmed.isEmpty()) return false
        if (trimmed.startsWith("{") || trimmed.startsWith("[") || trimmed.startsWith("\"") || trimmed.startsWith("-")) {
            return true
        }
        if (trimmed.first().isDigit()) {
            return true
        }
        return trimmed.startsWith("true") || trimmed.startsWith("false") || trimmed.startsWith("null")
    }

    private fun mapToJsonString(map: Map<String, Any?>): String {
        val element = anyToJsonElement(map)
        return json.encodeToString(JsonElement.serializer(), element)
    }

    @Suppress("UNCHECKED_CAST")
    private fun jsonStringToMap(str: String): Map<String, Any?> {
        val element = json.parseToJsonElement(str)
        return jsonElementToAny(element) as? Map<String, Any?> ?: emptyMap()
    }

    /** Close the underlying Ktor engine. Call from SDK destroy(). */
    fun close() {
        client.close()
    }

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun anyToJsonElement(value: Any?): JsonElement = when (value) {
            null -> JsonNull
            is Boolean -> JsonPrimitive(value)
            is Number -> JsonPrimitive(value)
            is String -> JsonPrimitive(value)
            is Map<*, *> -> JsonObject((value as Map<String, Any?>).mapValues { anyToJsonElement(it.value) })
            is List<*> -> JsonArray(value.map { anyToJsonElement(it) })
            else -> JsonPrimitive(value.toString())
        }

        fun jsonElementToAny(element: JsonElement): Any? = when (element) {
            is JsonNull -> null
            is JsonPrimitive -> {
                if (element.isString) element.content
                else element.longOrNull ?: element.doubleOrNull ?: element.booleanOrNull ?: element.content
            }
            is JsonObject -> element.entries.associate { (k, v) -> k to jsonElementToAny(v) }
            is JsonArray -> element.map { jsonElementToAny(it) }
        }
    }
}
